Ark panics during restore using DO plugin

**What steps did you take and what happened:**

Attempting to restore a backup of the example-nginx Pod & PV/PVC using the DigitalOcean plugin results in an Ark panic.

```
ark backup create namespace-only --include-namespaces=nginx-example
kubectl delete ns nginx-example --force --grace-period=0
ark restore create restore-ns --from-backup namespace-only
```

**What did you expect to happen:**

Should not panic - restore should complete.

**The output of the following commands will help us better understand what's going on**:
Logs can be found here: https://gist.github.com/rosskukulinski/d3655715348abc3d2ea6dca1033db418

**Environment:**
- Ark version (use `ark version`):
```
Version: v0.8.1
Git commit: v0.8.1
Git tree state: clean
```
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-08T16:31:10Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:43:26Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version:

StackPointCloud on DigitalOcean

- Cloud provider or hardware configuration:

Digital Ocean

- OS (e.g. from `/etc/os-release`):
CoreOS Stable

@skriss I can't seem to find where the sample plugin is loaded or is running.

I removed the DO plugin and am still getting the panic:

```
ark backup create default-no-do --include-namespaces=default
ark restore create testeroo --from-backup=default-no-do
```
logs:
```
time="2018-08-21T13:32:10Z" level=info msg="Starting backup" backup=heptio-ark/default-no-do logSource="pkg/controller/backup_controller.go:339"
time="2018-08-21T13:32:12Z" level=info msg="Backup completed" backup=heptio-ark/default-no-do logSource="pkg/controller/backup_controller.go:401"
time="2018-08-21T13:32:38Z" level=info msg="starting restore" backup=default-no-do logSource="pkg/controller/restore_controller.go:430" restore=heptio-ark/testeroo
E0821 13:32:40.663266       1 runtime.go:66] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/hashmap_fast.go:696
/go/src/github.com/heptio/ark/pkg/util/collections/map_utils.go:132
/go/src/github.com/heptio/ark/pkg/restore/merge_service_account.go:63
/go/src/github.com/heptio/ark/pkg/restore/restore.go:805
/go/src/github.com/heptio/ark/pkg/restore/restore.go:482
/go/src/github.com/heptio/ark/pkg/restore/restore.go:355
/go/src/github.com/heptio/ark/pkg/restore/restore.go:257
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:431
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:284
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:116
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:204
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:191
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:178
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:178
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map

goroutine 278 [running]:
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x17ff7a0, 0x1c3d310)
        /usr/local/go/src/runtime/panic.go:502 +0x229
github.com/heptio/ark/pkg/util/collections.MergeMaps(0x0, 0xc420a71f50)
        /go/src/github.com/heptio/ark/pkg/util/collections/map_utils.go:132 +0x11b
github.com/heptio/ark/pkg/restore.mergeServiceAccounts(0xc420a65480, 0xc420a65430, 0xf, 0x1a9d701, 0xc420a65430)
        /go/src/github.com/heptio/ark/pkg/restore/merge_service_account.go:63 +0x3df
github.com/heptio/ark/pkg/restore.(*context).restoreResource(0xc4206d3200, 0xc42039bb00, 0xf, 0xc42038fc34, 0x7, 0xc42038fc80, 0x3b, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/heptio/ark/pkg/restore/restore.go:805 +0x233b
github.com/heptio/ark/pkg/restore.(*context).restoreFromDir(0xc4206d3200, 0xc420a562c0, 0xe, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/heptio/ark/pkg/restore/restore.go:482 +0xd3d
github.com/heptio/ark/pkg/restore.(*context).execute(0xc4206d3200, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/heptio/ark/pkg/restore/restore.go:355 +0x451
github.com/heptio/ark/pkg/restore.(*kubernetesRestorer).Restore(0xc4200d24d0, 0xc420181680, 0xc4202cdd40, 0x1c41de0, 0xc42000c488, 0x1c41e00, 0xc42000c4c8, 0xc42020be00, 0x5, 0x8, ...)
        /go/src/github.com/heptio/ark/pkg/restore/restore.go:257 +0xbfc
github.com/heptio/ark/pkg/controller.(*restoreController).runRestore(0xc42031c410, 0xc420181680, 0xc4204ea3f0, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:431 +0xaa1
github.com/heptio/ark/pkg/controller.(*restoreController).processRestore(0xc42031c410, 0xc4203ef760, 0x13, 0x18, 0xc42067cdb8)
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:284 +0x705
github.com/heptio/ark/pkg/controller.(*restoreController).(github.com/heptio/ark/pkg/controller.processRestore)-fm(0xc4203ef760, 0x13, 0xc4204516c0, 0x16e1560)
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:116 +0x3e
github.com/heptio/ark/pkg/controller.(*restoreController).processNextWorkItem(0xc42031c410, 0xc42008a700)
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:204 +0xe0
github.com/heptio/ark/pkg/controller.(*restoreController).runWorker(0xc42031c410)
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:191 +0x2b
github.com/heptio/ark/pkg/controller.(*restoreController).(github.com/heptio/ark/pkg/controller.runWorker)-fm()
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:178 +0x2a
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4202d67a0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4208a7fa0, 0x3b9aca00, 0x0, 0x34630b8a001, 0xc4200af0e0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4202d67a0, 0x3b9aca00, 0xc4200af0e0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/heptio/ark/pkg/controller.(*restoreController).Run.func2(0xc42031c410, 0x1c621e0, 0xc42039e880, 0xc4204ea5f0)
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:178 +0x68
created by github.com/heptio/ark/pkg/controller.(*restoreController).Run
        /go/src/github.com/heptio/ark/pkg/controller/restore_controller.go:177 +0x31d
```

--- a/pkg/restore/merge_service_account.go
+++ b/pkg/restore/merge_service_account.go
@@ -58,9 +58,9 @@ func mergeServiceAccounts(fromCluster, fromBackup *unstructured.Unstructured) (*
 
 	desired.ImagePullSecrets = mergeLocalObjectReferenceSlices(desired.ImagePullSecrets, backupSA.ImagePullSecrets)
 
-	collections.MergeMaps(desired.Labels, backupSA.Labels)
+	desired.Labels = collections.MergeMaps(desired.Labels, backupSA.Labels)
 
-	collections.MergeMaps(desired.Annotations, backupSA.Annotations)
+	desired.Annotations = collections.MergeMaps(desired.Annotations, backupSA.Annotations)
 
 	desiredUnstructured, err := runtime.DefaultUnstructuredConverter.ToUnstructured(desired)
 	if err != nil {

--- a/pkg/util/collections/map_utils.go
+++ b/pkg/util/collections/map_utils.go
@@ -125,11 +125,18 @@ func Exists(root map[string]interface{}, path string) bool {
 
 // MergeMaps takes two map[string]string and merges missing keys from the second into the first.
 // If a key already exists, its value is not overwritten.
-func MergeMaps(first, second map[string]string) {
+func MergeMaps(first, second map[string]string) map[string]string {
+	// If the first map passed in is empty, just use all of the second map's data
+	if first == nil {
+		first = map[string]string{}
+	}
+
 	for k, v := range second {
 		_, ok := first[k]
 		if !ok {
 			first[k] = v
 		}
 	}
+
+	return first
 }

--- a/pkg/util/collections/map_utils_test.go
+++ b/pkg/util/collections/map_utils_test.go
@@ -18,6 +18,8 @@ package collections
 
 import (
 	"testing"
+
+	"github.com/stretchr/testify/assert"
 )
 
 func TestGetString(t *testing.T) {
@@ -42,5 +44,59 @@ func TestGetString(t *testing.T) {
 		if res != tc.result {
 			t.Error("res")
 		}
+	}
+}
+
+func TestMergeMaps(t *testing.T) {
+	var testCases = []struct {
+		name        string
+		source      map[string]string
+		destination map[string]string
+		expected    map[string]string
+	}{
+		{
+			name:        "nil destination should result in source being copied",
+			destination: nil,
+			source: map[string]string{
+				"k1": "v1",
+			},
+			expected: map[string]string{
+				"k1": "v1",
+			},
+		},
+		{
+			name: "keys missing from destination should be copied from source",
+			destination: map[string]string{
+				"k2": "v2",
+			},
+			source: map[string]string{
+				"k1": "v1",
+			},
+			expected: map[string]string{
+				"k1": "v1",
+				"k2": "v2",
+			},
+		},
+		{
+			name: "matching key should not have value copied from source",
+			destination: map[string]string{
+				"k1": "v1",
+			},
+			source: map[string]string{
+				"k1": "v2",
+			},
+			expected: map[string]string{
+				"k1": "v1",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			result := MergeMaps(tc.destination, tc.source)
+
+			assert.Equal(t, tc.expected, result)
+		})
 	}
 }

